### PR TITLE
New layout and #24 #22

### PR DESCRIPTION
**Pull request by Kasterisk**
Some adjustments to the layout, added selecting install location #22 and Intel GPU check to notify the player about the frequent crashes using the systeminformation node package #24 . Also added a new launcher option to hide the launcher on play

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<!DOCTYPE html>
+﻿<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply index.html]: Nothing is changed, I must've removed it by accident and then added it back in...

File: main.js [hunk 2]
```diff
@@ -5,7 +5,6 @@
 // use `npm run start-dev` if you want the dev tools
 // Set this to true if you want to open the dev console
 let openDev = false;
-
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply main.js]: Yeah, that also got removed somehow.

File: main.js [hunk 3]
```diff
@@ -75,13 +74,14 @@ function createWindow () {
...
-    
+
     // Create the browser window.
     mainWindow = new BrowserWindow({
-        width: 1200 + (openDev ? 700 : 0), height: 700,
+        width: 950 + (openDev ? 700 : 0), height: 600,
         // This would disable the system title bar and frame window
         // so if this is false we need a custom window top bar
         frame: true,
+        resizable: false,
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply main.js]: Currently the layout kinda breaks when resized, until its fixable i think the window should be unresizable temporarily

File: main.js [hunk 4]
```diff
@@ -195,5 +195,4 @@ electron.app.on('browser-window-created',function(e, window) {
     // Remove the menu bar with entries like "file" and "edit"
     if(!openDev)
         window.setMenu(null);
-});
-
+});
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply main.js]: will be fixed

File: modal.js [hunk 5]
```diff
@@ -89,7 +89,7 @@ class Modal {
 
         this.backdrop.style.display = "block";
         
-        $( this.dialog ).slideDown( 400, function() {
+        $( this.dialog ).slideDown( 500, function() {
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply modal.js]: will be readjusted

File: modal.js [hunk 6]
```diff
@@ -105,7 +105,7 @@ class Modal {
                 return;
         }
 
-        $( this.dialog ).slideUp( 400, () => {
+        $( this.dialog ).slideUp( 500, () => {
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply modal.js]: will also be readjusted

File: main.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (hhyyrylainen) [reply main.js]: That's acceptable. I'll just need to remember to create an issue for that if this isn't fixed before merging.

File: renderer.js [hunk 7]
```diff
@@ -985,14 +998,59 @@ function playPressed(){
 
         playModalQuitDLCancel = dataObj.reqObj;
         
+        
     });
 
 }
 
-playButtonText.addEventListener("click", function(event){
+async function checkIfCompatible() {
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply renderer.js]: yes, at the moment its just a basic intel checking. I will try to add those

File: renderer.js [hunk 8]
```diff
@@ -1066,7 +1124,7 @@ function updatePlayButtonText(){
 
     assert(download);
     
-    playButtonText.textContent = "Play " + version.getDescriptionString() + " " +
+    playButtonText.textContent = "Play " + version.getDescriptionString() + "\n" +
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply renderer.js]: some dumb mistake that i forgot to remove, sorry about that one lol

File: renderer.js [hunk 9]
```diff
@@ -985,14 +998,59 @@ function playPressed(){
...
+        });
+    
+        closeContainer.append(close);
+        status.append(closeContainer);
+    }
+    else{
+        playPressed();
+
+        if(settings.hideLauncherOnPlay){
+            win.hide();
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply renderer.js]: alright, that will be changed

**Comment by Kasterisk** [comment]
That's weird, I've found the opposite to be true, the 0.4.2 version always opened up instead.

Also, I have managed to add a dialog that asks the player if the installed files are to be automatically moved to the selected new directory, so that will be very convenient. And a reset button will be nice

File: index.html [hunk 10]
```diff
@@ -142,18 +146,35 @@ <h2>News</h2>
...
                 Browse Installed Files
               </p>
             </div>
+            <div id="selectInstallLocation" class="BottomButton" style="height: 40px;">
+              <p style="text-align: left; font-size: 0.8em;">
+                Select Install Location
+              </p>
+            </div>
+            <div id="resetInstallLocation" class="BottomButton" style="font-size: 1.5em; vertical-align: middle;">
+                &#x21bb
```
Comment (hhyyrylainen) [inline index.html]: I think this needs a tooltip or something. To me just the icon isn't super clear what it does.

File: install_handler.js [hunk 11]
```diff
@@ -76,6 +77,28 @@ function deleteInstalledVersion(name){
...
+                    }
+                });
+            }
+        }
+    });
+}
 
 module.exports.listInstalledVersions = listInstalledVersions;
 module.exports.deleteInstalledVersion = deleteInstalledVersion;
+module.exports.moveInstalledVersion = moveInstalledVersion;
```
Comment (hhyyrylainen) [inline install_handler.js]: missing newline

File: renderer.js [hunk 12]
```diff
@@ -31,10 +34,17 @@ var pjson = require('./package.json');
...
+const { settings, loadSettings, saveSettings, dataFolder, tmpDLFolder,
+        locallyCachedDLFile, getInstallPath, insDirs, saveInstalledDir } =
       require('./settings.js');
 
+checkIfCompatible();
+
+let graphicControllers = null;
+let cards = {};
+
+let showIncompatiblePopup = false;
```
Comment (hhyyrylainen) [inline renderer.js]: Shouldn't these variables be defined *before* the call to checkIfCompatible?

File: renderer.js [hunk 13]
```diff
@@ -985,14 +1007,72 @@ function playPressed(){
...
+
+    if(showIncompatiblePopup){
+        status.textContent = "Detected graphics card(s): " + cards;
+        status.append(document.createElement("br"));
+        status.append(document.createElement("br"));
+        status.append(document.createTextNode("WARNING: Intel Integrated Graphics Card may causes Thrive to crash due to graphics engine issues with Intel: https://github.com/Revolutionary-Games/Thrive/issues/804."))
+        status.append(document.createElement("br"));
+        status.append(document.createTextNode("This is a known problem, any help fixing this would be most welcome!"));
+
+        let closeContainer = document.createElement("div");
```
Comment (hhyyrylainen) [inline renderer.js]: If the detected graphics has "nvidia" or "amd" in it an additional help text telling the user to configure thrive to run with that graphics card, would be helpful.

File: renderer.js [hunk 14]
```diff
@@ -1066,7 +1146,7 @@ function updatePlayButtonText(){
 
     assert(download);
     
-    playButtonText.textContent = "Play " + version.getDescriptionString() + " " +
+    playButtonText.textContent = "Play " + version.getDescriptionString() +
```
Comment (hhyyrylainen) [inline renderer.js]: The space here should be kept.

File: settings.js [hunk 15]
```diff
@@ -9,48 +9,66 @@ const mkdirp = require('mkdirp');
...
+    hideLauncherOnPlay: true,
+    installDir: path.join(module.exports.dataFolder, "Installed"),
 };
 
+module.exports.insDirs = {
+    installedDir: null,
+}
+
 const settingsFile = path.join(module.exports.dataFolder, "launcher_settings.json");
+const installedDirFile = path.join(module.exports.dataFolder, "installed_directories.json");
```
Comment (hhyyrylainen) [inline settings.js]: Why is this a separate file?
Wouldn't it make more sense to have this in the main settings file? The default path could be renamed to `defaultInstallDir` to free up the name for the actual current install directory.

File: install_handler.js [hunk 16]
```diff
@@ -21,7 +22,7 @@ function listInstalledVersions(){
 
         const versions = getVersionData().versions;
 
-        const directories = getDirectories(installPath);
+        const directories = getDirectories(insDirs.installedDir);
```
Comment (hhyyrylainen) [inline install_handler.js]: This function causes an error for me:
```
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be one of type string, Buffer, or URL. Received type object
```
At least I think that comes from here. That's in the space where the installed versions should be listed.

File: settings_dialog.js [hunk 17]
```diff
@@ -101,14 +107,70 @@ function onSettingsChanged(){
...
+            currentInstallDir.innerHTML = "Directory: " + getInstallPath();
+
+            installedMessageBox();
+        }
+    });
+});
+
+let resestInstallLocation = document.getElementById("resetInstallLocation");
+
+resetInstallLocation.addEventListener("click", function(event){
```
Comment (hhyyrylainen) [inline settings_dialog.js]: This function should not do anything (and the button could be disabled) if the current install location is the default one.

**Comment by Kasterisk** [comment]
Ok i will look into this bug. Also, do i need to change the scrollbar back (to the new one)?

File: main.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kasterisk) [reply main.js]: I will try to work on this

**Comment by hhyyrylainen** [comment]
Changing the scroll bar back to the new one would be nice.

**Comment by Kasterisk** [comment]
The Browse Installed Files button actually opens up the installed directory, and the installedDir are only used as a path reference for moving the files from where they are to installDir and then after moving the files done, its value are set as same as installDir. And I think in your case the default install path are empty so the launcher doesn't actually move any files after changing the install location.

**Comment by Kasterisk** [comment]
I've made a commit so that the files will always be moved after the install location are changed. Also, checks if the installedDir are empty, if so then set the value of installedDir the same as installDir anyway.

**Comment by Kasterisk** [comment]
Yeah, turns out installedDir is the obvious problem. I've made some changes to how the launcher moves the files, seems to work fine with me though, don't know if its gonna cause more problems

File: renderer.js [hunk 18]
```diff
@@ -985,14 +1010,85 @@ function playPressed(){
...
+        closeContainer.style.marginTop = "20px";
+        closeContainer.style.textAlign = "center";
+        let close = document.createElement("div");
+        close.textContent = "Continue";
+        close.className = "CloseButton";
+    
+        close.addEventListener('click', (event) => {
+            incompatibleModal.hide();
+            if(cards != null){
+                playPressed();
```
Comment (hhyyrylainen) [inline renderer.js]: Maybe `showIncompatiblePopup` should be set to false here to only show the box once?
We can later add an option to permanently dismiss the popup.

File: renderer.js [hunk 19]
```diff
@@ -985,14 +1010,85 @@ function playPressed(){
...
+            }
+        });
+
+        closeContainer.append(close);
+        box.append(closeContainer);
+        showIncompatiblePopup = false;
+    }
+    else{
+        if(cards != null){
```
Comment (hhyyrylainen) [inline renderer.js]: I think that the game should be started even if the card detection fails, as I'm not sure how bulletproof that is.

File: settings.js [hunk 20]
```diff
@@ -9,31 +9,41 @@ const mkdirp = require('mkdirp');
...
                                        "Revolutionary-Games-Launcher");
 
-module.exports.installPath = path.join(module.exports.dataFolder, "Installed");
-
 module.exports.locallyCachedDLFile = path.join(module.exports.dataFolder,
                                                "saved_version_db_v2.json");
 
+module.exports.defaultInstallPath = path.join(module.exports.dataFolder, "Installed");
+
+module.exports.setInstallPath = (directory) => {
```
Comment (hhyyrylainen) [inline settings.js]: I still think this property should be just added to the main settings object, and have no separate get or set functions.

Comment (Kasterisk) [reply settings.js]: Nevermind, I got it.

File: renderer.js [hunk 21]
```diff
@@ -985,14 +1010,84 @@ function playPressed(){
...
+        // Get all of the graphics cards info
+        var i;
+        for (i = 0; i < data.controllers.length; i++){
+            graphicControllers = data.controllers[i];
+            cards = String(graphicControllers.model);
+            console.log(cards);
+        }
+
+        // Is incompatible if the word Intel is found in a substring
+        if(cards.includes("Intel")){
```
Comment (hhyyrylainen) [inline renderer.js]: This does not seem to work. I tested on my laptop that has nvidia and intel graphics, and it did not give me any kind of popup.

File: settings_dialog.js [hunk 22]
```diff
@@ -79,14 +98,73 @@ function updateInstalledVersions(){
     });    
 }
 
+function checkForInstalledVersions(){
```
Comment (hhyyrylainen) [inline settings_dialog.js]: Why is this needed? There is already `listInstalledVersions` in `install_handler.js`

File: renderer.js [hunk 23]
```diff
@@ -985,14 +1011,92 @@ function playPressed(){
...
+    try {
+        const data = await si.graphics();
+
+        // Get all of the graphics cards info
+        for (var i = 0; i < data.controllers.length; i++){
+            graphicControllers = data.controllers[i];
+            cardsVendor = String(graphicControllers.vendor.toLowerCase());
+            cardsModel = String(graphicControllers.model + ", ");
+            console.log(graphicControllers.model);
+        }
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply renderer.js]: I tested it on an amd card laptop, and the gpu vendor info actually does says Advanced Micro Devices, Inc. But, I will add it in just in case though...

File: settings_dialog.js [hunk 24]
```diff
@@ -1,41 +1,60 @@
...
 
 const settingsModal = new Modal("settingsModal", "settingsModalDialog", {
     closeButton: "settingsClose"
 });
 
+const movingFileModal = new Modal("movingFileModal", "movingFileModalDialog", {
+    autoClose: false
+});
+
+let installedVersions = [];
```
Comment (hhyyrylainen) [inline settings_dialog.js]: I think that this global variable could be avoided by just calling the version finding code whenever the list is needed (and/or passing it around as a parameter).

File: renderer.js (hunk 18 above)
Comment (hhyyrylainen) [reply renderer.js]: This still seems to be the case...

File: renderer.js [hunk 25]
```diff
@@ -355,7 +367,45 @@ function onVersionDataReceived(data, unsigned = false){
...
 }
 
-function loadVersionData(){
+// Checks the graphics card
+async function checkIfCompatible() {
+    try {
+        const data = await si.graphics();
+        const identifier = ["nvidia", "advanced micro devices", "amd"]; // and so on...
+
+        let cardsVendor = [];
```
Comment (hhyyrylainen) [inline renderer.js]: Is this variable needed?

File: renderer.js [hunk 26]
```diff
@@ -355,7 +367,45 @@ function onVersionDataReceived(data, unsigned = false){
...
+
+        for(var i = 0; i < data.controllers.length; i++){
```
Comment (hhyyrylainen) [inline renderer.js]: The vars in these loops should be `let`

File: renderer.js [hunk 27]
```diff
@@ -355,7 +367,45 @@ function onVersionDataReceived(data, unsigned = false){
...
+			
+            let graphicControllers = data.controllers[i];
+			
+            cardsVendor.push(graphicControllers.vendor.toLowerCase());
+            cardsModel.push(" " + graphicControllers.model);
+            
+            for(var n = 0; n < cardsVendor.length; n++){
```
Comment (hhyyrylainen) [inline renderer.js]: This loop is totally unnecessary as each vendor / card is already looped through by the top level loop (starting at line 378).

**Comment by hhyyrylainen** [comment]
I tested the logic again and it now works. Two comments about the incompatible popup. The warning text should be the topmost thing and then the list of graphics cards and the help text for using a different card.
Also the link to the issue should be an actual clickable link.

File: settings_dialog.js (hunk 24 above)
Comment (Kasterisk) [reply settings_dialog.js]: May I just change it to a local variable?

Comment (hhyyrylainen) [reply settings_dialog.js]: Yes, you can do that.

**Comment by Kasterisk** [comment]
Alright thanks, I will try to fix it. Also, it seems the `checkIfCompatible` function could be delayed causing the incompatible popup not to show up if the user is quickly pressing play right after opening up the launcher.

**Comment by hhyyrylainen** [comment]
I would assume that the time it takes to query what hardware there is, is just a couple of milliseconds. Though, I didn't measure so I can't be sure, but I would think we can assume it always has time to run before the user manages to click anything.

**Comment by Kasterisk** [comment]
Ok, I've made an another commit. Hopefully everything works fine now.

File: settings_dialog.js [hunk 28]
```diff
@@ -104,11 +146,88 @@ function onSettingsChanged(){
...
 browseFilesButton.addEventListener("click", function(event){
-    const target = installPath;
+    const target = settings.installPath;
     console.log("Opening item:", target);
     shell.openItem(target);
 });
 
+function askToMoveFiles(selectedDirectory){
+
+    let files = [];
```
Comment (hhyyrylainen) [inline settings_dialog.js]: This variable should be defined inside the callback. This is filled by that asynchronous operation, it makes no sense that this is outside that scope.

File: settings_dialog.js [hunk 29]
```diff
@@ -79,14 +94,41 @@ function updateInstalledVersions(){
     });    
 }
 
+function moveInstalledFiles(filePath, destination){
```
Comment (hhyyrylainen) [inline settings_dialog.js]: `filePath` should be named `files` as it is a list of files and not a path.

File: settings_dialog.js [hunk 30]
```diff
@@ -79,14 +94,41 @@ function updateInstalledVersions(){
...
+    
+    for (let i = 0; i < filePath.length; i++) {
+        let file = filePath[i];
+        let fileName = path.basename(filePath[i]);
+
+        movingFileModal.show();
```
Comment (hhyyrylainen) [inline settings_dialog.js]: If this content doesn't change per loop iteration, it should be outside the loop. Also this way having it constantly pop up and close is not very good.

File: settings_dialog.js [hunk 31]
```diff
@@ -79,14 +94,41 @@ function updateInstalledVersions(){
...
+        content.append(document.createTextNode("This may take several minutes, please be patient."))
+        
+        fs.move(file, path.join(destination, fileName))
+        .then(() => {
+            console.log("Successfully moving: " + fileName);
+
+            settings.installPath = destination;
+            onSettingsChanged();
+            updateInstalledVersions();
+            movingFileModal.hide();
```
Comment (hhyyrylainen) [inline settings_dialog.js]: This happens for each successfully moved file? Shouldn't this instead happen after moving all the files?

File: settings_dialog.js [hunk 32]
```diff
@@ -1,41 +1,55 @@
...
 });
 
+const movingFileModal = new Modal("movingFileModal", "movingFileModalDialog", {
+    autoClose: false
+});
+
 // Used to skip callbacks on loading settings
 let loadingSettings = false;
 
+let listInstalledVersionsError = false;
```
Comment (hhyyrylainen) [inline settings_dialog.js]: This variable is not read anywhere (used for anything) and can be removed.

File: settings.js [hunk 33]
```diff
@@ -9,24 +9,26 @@ const mkdirp = require('mkdirp');
...
-
 module.exports.locallyCachedDLFile = path.join(module.exports.dataFolder,
                                                "saved_version_db_v2.json");
 
+module.exports.defaultInstallPath = path.join(module.exports.dataFolder, "Installed");
 
 // Make sure it exists. This simplifies a lot of code
 mkdirp.sync(module.exports.dataFolder);
 
+mkdirp.sync(module.exports.defaultInstallPath);
```
Comment (hhyyrylainen) [inline settings.js]: The user likely won't appreciate unnecessary folders being created, especially if they change the install path.
Code that breaks without the install path existing should catch that as an error (other than the version downloading code which should be the only )

Comment (Kasterisk) [reply settings.js]: Hmm... I thought that was a bug.

File: index.html [hunk 34]
```diff
@@ -13,44 +13,50 @@
...
+      style="height: 5%; width: calc(35% + 5px); bottom: calc(70px + 22px); left: 8px;">
+    </span>
+
     <span class="ContentBlock"
-          style="height: 90%; width: 40%; top: 0px; left: 0px; border-right: 1px solid 
+          style="height: calc(80% + 10px); width: calc(35% + 5px); top: 10px; left: 8px;
                  rgba(68, 68, 68, 0.46)">
 
-      <h2>Development Forums</h2>
+      <h2>DEVELOPMENT FORUMS</h2>
```
Comment (Themayu) [inline index.html]: What's with the capitalisation change here?

File: index.html [hunk 35]
```diff
@@ -13,44 +13,50 @@
...
+    <span class="BottomFade"
+      style="height: 5%; width: calc(60% + 12px); bottom: calc(70px + 22px); left: calc(35% + 25px);">
+    </span>
+    
     <span class="ContentBlock"
-          style="height: 90%; width: calc(60% - 1px); top: 0px; left: calc(40% + 1px);">
+          style="height: calc(80% + 10px); width: calc(60% + 12px); top: 10px; left: calc(35% + 25px)">
 
-      <h2>News</h2>
+      <h2 style="padding-left: 50px;">NEWS</h2>
```
Comment (Themayu) [inline index.html]: Another capitalisation change that I find myself perplexed by, along with the indentation. In my opinion, this doesn't look very good when viewed in the launcher. Why?

File: main.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Themayu) [reply main.js]: I think it would be nice to always allow resizing when in development mode, as the hard-coded value here makes it hard to use the Chrome Developer Tools for debugging the UI.

File: settings_dialog.js (hunk 31 above)
Comment (Kasterisk) [reply settings_dialog.js]: Well, it should. Could I check if all the files have been moved using `if(i == files.length - 1)` inside the loop?

File: package.json [hunk 36]
```diff
@@ -82,14 +82,16 @@
     },
     "dependencies": {
         "feedparser": "^2.2.9",
+        "fs-extra": "^8.1.0",
```
Comment (Themayu) [inline package.json]: What's the use of `fs-extra` here? Because at the moment, it looks like we're adding a 47.14kb package to the launcher just to promisify a single method - which is what the built-in `util.promisify` function is for.

File: settings_dialog.js (hunk 32 above)
Comment (Themayu) [reply settings_dialog.js]: I agree with Hhyyrylainen here. The variable seems redundant in its current state, and as it is just creates unneeded noise in the project. I say either make use of it, or strip it out.

File: install_handler.js [hunk 37]
```diff
@@ -2,26 +2,27 @@
 // Helpers for installing and removing versions
 //
 // TODO: the installing functions are still in renderer.js and should be moved here
-const fs = require('fs');
+const fs = require('fs-extra');
```
Comment (Themayu) [inline install_handler.js]: The use of `fs-extra` here is unnecessary, as all methods used from it are available in `fs`.

File: settings_dialog.js (hunk 31 above)
Comment (hhyyrylainen) [reply settings_dialog.js]: Then you would know if the *last* file succeeded or not. And it's not guaranteed that the order is the same. What I'd do is chain the returned promises from the file move. Or just await for each promise to succeed before moving the next file.

File: settings_dialog.js [hunk 38]
```diff
@@ -1,41 +1,55 @@
 // Everything under the settings button
 "use strict";
 
+const fs = require('fs-extra');
```
Comment (Themayu) [inline settings_dialog.js]: `fs-extra` is being imported to use a single promisified version of a method available in `fs`. I recommend removing `fs-extra` from the package and simply using `util.promisify` to transform the method that is used.

File: settings_dialog.js (hunk 31 above)
Comment (Themayu) [reply settings_dialog.js]: `fs.move` returns after all files have been moved; not for every file. Promises can only `resolve` or `reject` once.

File: settings_dialog.js [hunk 39]
```diff
@@ -1,41 +1,55 @@
...
 const path = require('path');
 
 var {shell} = require('electron');
+const {dialog} = require('electron').remote;
+const win = remote.getCurrentWindow();
+
+const {getVersionData} = require("./version_info.js");
```
Comment (Themayu) [inline settings_dialog.js]: Why is `getVersionData` being retrieved, if it is not used?

File: settings_dialog.js [hunk 40]
```diff
@@ -1,41 +1,55 @@
...
 
 const { Modal, showGenericError} = require('./modal');
-const {listInstalledVersions, deleteInstalledVersion} = require('./install_handler.js');
-const { settings, saveSettings, installPath } = require('./settings.js');
+const { listInstalledVersions, deleteInstalledVersion } = require('./install_handler.js');
+
+const { settings, dataFolder, saveSettings, defaultInstallPath } = require('./settings.js');
```
Comment (Themayu) [inline settings_dialog.js]: Again with `dataFolder` here. It is not being used, so why import it?

**Comment by Themayu** [comment]
Zuris#4495 here; launcher developer (even though I haven't made a single PR yet. Still working on it, though!)

I've pointed out a few extra problems I can find with the current state of the Pull Request, to go alongside the issues Hhyyrylainen has found.

File: settings_dialog.js (hunk 31 above)
Comment (Themayu) [reply settings_dialog.js]: Ah, wait, misread the code itself. Ignore me.

File: settings_dialog.js (hunk 32 above)
Comment (Kasterisk) [reply settings_dialog.js]: True, that will be removed.

Comment (Kasterisk) [reply settings_dialog.js]: True, that will be removed.

File: index.html (hunk 34 above)
Comment (Kasterisk) [reply index.html]: I though that will make the box title to be more readable.

Comment (hhyyrylainen) [reply index.html]: I don't particularly mind either way. When I posted a test build for other developers no one else seemed to mind either, so... I think this change can stay.

File: index.html (hunk 35 above)
Comment (Kasterisk) [reply index.html]: Also the same reason as before. The original uncapitalized one looks fine though, i could change to that.

File: settings_dialog.js (hunk 40 above)
Comment (Kasterisk) [reply settings_dialog.js]: Didn't notice that one before, that will be removed.

File: package.json (hunk 36 above)
Comment (Kasterisk) [reply package.json]: Currently, I'm using it for the `fs.move` method. As I'm not really sure if `fs.rename` works well, is it a decent option though?

Comment (hhyyrylainen) [reply package.json]: I didn't look into this a ton, but it seems that fs-extra works even between different partitions: https://stackoverflow.com/questions/4568689/how-do-i-move-file-a-to-a-different-partition-or-device-in-node-js
whereas plain fs whould need bunch of extra code to handle that. So fs-extra use should be kept.

File: settings_dialog.js (hunk 31 above)
Comment (Kasterisk) [reply settings_dialog.js]: This should be fixed as of the latest commit.

File: settings_dialog.js [hunk 41]
```diff
@@ -79,14 +90,39 @@ function updateInstalledVersions(){
...
+    content.innerHTML = "Moving files to: " + destination + " ...";
+    content.append(document.createElement("br"));
+    content.append(document.createTextNode("This may take several minutes, please be patient."));
+
+    await Promise.all(files.map(file =>
+        fs.move(file, path.join(destination, path.basename(file)))
+        .then(() => {
+            console.log("finished moving: " + path.basename(file));
+        })
+        .catch (err => {
```
Comment (Kasterisk) [reply settings_dialog.js]: Yeah, that will be much more better.

File: install_handler.js (hunk 37 above)
Comment (hhyyrylainen) [reply install_handler.js]: Is this still the case? Even if fs-extra is needed the variable it is assigned to should be `fsExtra` to be clear that it isn't `fs`

File: settings_dialog.js [hunk 42]
```diff
@@ -71,6 +81,7 @@ function updateInstalledVersions(){
         }
 
     }).catch(err => {
+        listInstalledVersionsError = true;
```
Comment (hhyyrylainen) [inline settings_dialog.js]: This variable was removed, but is still referenced here.

**Comment by Kasterisk** [comment]
Should the `checkIfCompatible` be called only once? I think that will reduce the launcher's initialization time quite a bit after the first time use.

**Comment by hhyyrylainen** [comment]
`checkIfCompatible` should only be called once

File: settings_dialog.js (hunk 41 above)
Comment (hhyyrylainen) [inline settings_dialog.js]: Wouldn't it work better if these .then and .catch would be moved to the Promise.all so that the first error of all of them is detected as well as the success is detected when all succeed.
Right now I think it's possible to trigger both the error and hit the successfully moved line.

File: settings_dialog.js [hunk 43]
```diff
@@ -79,14 +90,39 @@ function updateInstalledVersions(){
...
+    settings.installPath = destination;
+    onSettingsChanged();
+    updateInstalledVersions();
+    movingFileModal.hide();
+}
+
 settingsButton.addEventListener("click", function(event){
 
-    settingsModal.show();
+    settingsModal.show();
```
Comment (hhyyrylainen) [inline settings_dialog.js]: Extra whitespace added here.

File: renderer.js (hunk 18 above)
Comment (hhyyrylainen) [reply renderer.js]: Still not fixed.

**Comment by Kasterisk** [comment]
Ignore those same commits title...

File: settings_dialog.js [hunk 44]
```diff
@@ -79,14 +89,44 @@ function updateInstalledVersions(){
...
+                onSettingsChanged();
+                updateInstalledVersions();
+                movingFileModal.hide();
+            })
+            .catch(err => {
+                movingFileModal.hide();
+                showGenericError("Failed to move file(s): " + err.message);
+            });
+
+    } catch (err) {
```
[the remark this answers is not in the document]
Comment (Kasterisk) [reply settings_dialog.js]: Just in case something went wrong before moving the files, but yeah this is probably redundant.

**Comment by hhyyrylainen** [comment]
https://github.com/Revolutionary-Games/Thrive-Launcher/issues/38
